# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxPlatformFontList.h
# Commit: f44f01e127f0
# Full Hash: f44f01e127f0dcc96d74bd10606b318e41946906
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2021-05-22 03:53:20
# Description:
#   Bug 1712165 - Reorder conditions to avoid a possible race between the InitFontList thread and main thread accessing gfxPlatformFontList. r=lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D115690
# ==============================================================================

diff -r 360f7d0ede58 -r f44f01e127f0 gfx/thebes/gfxPlatformFontList.h
--- a/gfx/thebes/gfxPlatformFontList.h	Fri May 21 16:30:21 2021 +0000
+++ b/gfx/thebes/gfxPlatformFontList.h	Fri May 21 16:39:17 2021 +0000
@@ -170,12 +170,14 @@
   typedef nsTArray<FontFamily> PrefFontList;
 
   static gfxPlatformFontList* PlatformFontList() {
-    if (sPlatformFontList->IsInitialized()) {
-      return sPlatformFontList;
-    }
-    // Currently, only macOS uses OMT font-list initialization; on other
-    // platforms we initialize it directly during gfxPlatform::Init().
+    // If there is a font-list initialization thread, we need to let it run
+    // to completion before the font list can be used for anything else.
     if (sInitFontListThread) {
+      // If we're currently on the initialization thread, just continue;
+      // otherwise wait for it to finish.
+      if (IsInitFontListThread()) {
+        return sPlatformFontList;
+      }
       PR_JoinThread(sInitFontListThread);
       sInitFontListThread = nullptr;
       // If font-list initialization failed, the thread will have cleared
@@ -185,8 +187,10 @@
         MOZ_CRASH("Could not initialize gfxPlatformFontList");
       }
     }
-    if (!sPlatformFontList->InitFontList()) {
-      MOZ_CRASH("Could not initialize gfxPlatformFontList");
+    if (!sPlatformFontList->IsInitialized()) {
+      if (!sPlatformFontList->InitFontList()) {
+        MOZ_CRASH("Could not initialize gfxPlatformFontList");
+      }
     }
     return sPlatformFontList;
   }
